### PR TITLE
Fix #8985 'describe-security-group-rules' to return rules description

### LINKED ISSUE #8985: bug: `describe-security-group-rules` not returning rules description
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Command `describe-security-group-rules` does not return rules description.

### Expected Behavior

When a rule description is present, it should be included among the fields returned by `describe-security-group-rules`.

### How are you starting LocalStack?

With a docker-compose file

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

```yaml
version: "3.7"
services:
  aws:
    image: localstack/localstack:2.2
    environment:
      SERVICES: "ec2"
```

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

```sh
# Create a security group
security_group_id=$(awslocal ec2 create-security-group --group-name example --description "Example security group" | jq -r .GroupId)

# Add a rule with a description
awslocal ec2 authorize-security-group-ingress --group-id "${security_group_id}" --ip-permissions 'IpProtocol=tcp,FromPort=3389,ToPort=3389,IpRanges=[{CidrIp=127.0.0.1/32,Description=test-description}]'

# Optionally make sure the description was set (describe-security-groups behaves correctly)
#     awslocal ec2 describe-security-groups --group-ids "${security_group_id}"

# 🐞 Rules are returned but the description field is missing
awslocal ec2 describe-security-group-rules --filters "Name=group-id,Values=${security_group_id}"
```

### Environment

```markdown
- OS: Ubuntu 22.04 (ARM64)
- LocalStack: 2.2.0 (build date: 2023-07-20, git hash: 2d70f768, repo digest: localstack/localstack@sha256:5ff0dfc3b2c7c84a0bef449ff144e523b28bfc0e7f6a4304ac3b41d956eb6920)
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by ashish1500616**
Fixes #8985 

Label : https://github.com/localstack/localstack/labels/semver%3A%20patch

# Motivation.

As reported in #8985 ,  ```describe-security-group-rules``` command does not return rules description.

The Moto backend was returning an object of SecurityGroupRule that contained various attributes, including PrefixListId, GroupId, and CidrIp, as illustrated in the examples below:

```BASH
[{'PrefixListId': 'pl-78a54011', 'Description': 'pl-78a54011-description'}]
[{'GroupId': 'sg-f7da0aeb67db76223', 'Description': 'Allow traffic from another security group', 'OwnerId': '000000000000'}]
[{'CidrIp': '127.0.0.1/32', 'Description': 'test-description1'}, {'CidrIpv6': '2001:0db8:85a3:0000:0000:8a2e:0370:7334/128', 'Description': 'IPv6 test description'}]
```

According to my analysis, the ```moto.ec2.responses.DESCRIBE_SECURITY_GROUP_RULES_RESPONSE``` does not currently provide support for parsing responses when a SecurityGroupRule object contains fields such as PrefixListId, GroupId, and CidrIpv6.

# Changes

Implemented modifications in the pull request to align the response with AWS standards and return ```Description```

## Before
```BASH
security_group_id=$(awslocal ec2 create-security-group --group-name example --description "Test Security Group" | jq -r .GroupId)
```
```BASH
awslocal ec2 authorize-security-group-ingress --group-id "${security_group_id}" --ip-permissions 'IpProtocol=tcp,FromPort=3389,ToPort=3389,IpRanges=[{CidrIp=127.0.0.1/32,Description=test-description1}],PrefixListIds=[{PrefixListId=pl-78a54011,Description=pl-78a54011-description}],Ipv6Ranges=[{CidrIpv6=2001:0db8:85a3:0000:0000:8a2e:0370:7334/128,Description=IPv6 test description}],UserIdGroupPairs=[{GroupId='${security_group_id}',Description=Allow traffic from another security group}]'
```
```BASH
awslocal ec2 describe-security-group-rules --filters "Name=group-id,Values=${security_group_id}" 
{
    "SecurityGroupRules": [
        {
            "SecurityGroupRuleId": "sgr-c67a52ca18e6cdac2",
            "GroupId": "sg-cdc859e438b01a8e9",
            "GroupOwnerId": "000000000000",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "CidrIpv4": "127.0.0.1/32"
        },
        {
            "SecurityGroupRuleId": "sgr-e7cbce7cd2fa11415",
            "GroupId": "sg-cdc859e438b01a8e9",
            "GroupOwnerId": "000000000000",
            "IsEgress": true,
            "IpProtocol": "-1",
            "CidrIpv4": "0.0.0.0/0"
        }
    ]
}
```
## After
```BASH
awslocal ec2 authorize-security-group-ingress --group-id "${security_group_id}" --ip-permissions 'IpProtocol=tcp,FromPort=3389,ToPort=3389,IpRanges=[{CidrIp=127.0.0.1/32,Description=test-description1}],PrefixListIds=[{PrefixListId=pl-78a54011,Description=pl-78a54011-description}],Ipv6Ranges=[{CidrIpv6=2001:0db8:85a3:0000:0000:8a2e:0370:7334/128,Description=IPv6 test description}],UserIdGroupPairs=[{GroupId='${security_group_id}',Description=Allow traffic from another security group}]'
{
    "Return": true,
    "SecurityGroupRules": [
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "GroupOwnerId": "000000000000",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "CidrIpv4": "127.0.0.1/32",
            "Description": "test-description1"
        },
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "GroupOwnerId": "000000000000",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "CidrIpv6": "2001:0db8:85a3:0000:0000:8a2e:0370:7334/128",
            "Description": "IPv6 test description"
        },
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "GroupOwnerId": "000000000000",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "PrefixListId": "pl-78a54011",
            "Description": "pl-78a54011-description"
        },
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "GroupOwnerId": "000000000000",
            "IsEgress": true,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "ReferencedGroupInfo": {
                "GroupId": "sg-c8a89d730e37ac39e",
                "UserId": "000000000000"
            },
            "Description": "Allow traffic from another security group"
        }
    ]
}
```
```BASH
awslocal ec2 describe-security-group-rules --filters "Name=group-id,Values=${security_group_id}"
{
    "SecurityGroupRules": [
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "CidrIpv4": "127.0.0.1/32",
            "Description": "test-description1"
        },
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "CidrIpv6": "2001:0db8:85a3:0000:0000:8a2e:0370:7334/128",
            "Description": "IPv6 test description"
        },
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "ReferencedGroupInfo": {
                "GroupId": "sg-c8a89d730e37ac39e",
                "UserId": "000000000000"
            },
            "Description": "Allow traffic from another security group"
        },
        {
            "SecurityGroupRuleId": "sgr-e714ef9d4dab5e113",
            "GroupId": "sg-c8a89d730e37ac39e",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 3389,
            "ToPort": 3389,
            "PrefixListId": "pl-78a54011",
            "Description": "pl-78a54011-description"
        },
        {
            "SecurityGroupRuleId": "sgr-bb0c717b7d5657733",
            "GroupId": "sg-c8a89d730e37ac39e",
            "IsEgress": true,
            "IpProtocol": "-1",
            "CidrIpv4": "0.0.0.0/0"
        }
    ]
}
```

### CODE REVIEW

**Comment by ashish1500616** [comment]
Hi @viren-nadkarni 

I hope you're doing well. I've made some changes to fix #9059 and I would greatly appreciate your review.

Thanks

File: localstack/services/ec2/exceptions.py [hunk 1]
```diff
@@ -54,3 +56,21 @@ def __init__(self):
...
+        )
+
+
+class InvalidGroupIdMalformed(CommonServiceException):
+    def __init__(
+        self,
+    ):
+        super().__init__(
+            "InvalidGroupId.Malformed",
+            "The security group ID '' is malformed",
```
Comment (viren-nadkarni) [inline localstack/services/ec2/exceptions.py:75]: the security group ID is omitted: `''`

**Review by viren-nadkarni — CHANGES_REQUESTED** [review]
Thanks for the PR. This seems like quite a bit of logic for filtering. Have you taken a look at Moto filter helpers? Could you also please add some tests for this?

File: localstack/services/ec2/exceptions.py (hunk 1 above)
Comment (ashish1500616) [reply localstack/services/ec2/exceptions.py:75]: Hey @viren-nadkarni ,

When I invoke the ```describe-security-group-rules``` command with an empty group-id value on AWS it returns an error message ```InvalidGroupId.Malformed```

```bash
aws ec2 describe-security-group-rules --filters "Name=group-id,Values=['']"
An error occurred (InvalidGroupId.Malformed) when calling the DescribeSecurityGroupRules operation: The security group ID '' is malformed
```
And when I use the same on localstack,it also returns an error, but with a different message:
```bash
awslocal ec2 describe-security-group-rules --filters "Name=group-id,Values=['']"
An error occurred (InvalidGroup.NotFound) when calling the DescribeSecurityGroupRules operation: The security group 'No security groups found matching the filters provided.' does not exist
```

It appears that Moto is returning InvalidGroup.NotFound for both cases when the group-id value is empty and when the security group doesn't exist. 
Could you please confirm if this behaviour in Moto is intentional?
If that's the case, I'm willing to make the necessary adjustments.

**Comment by ashish1500616** [comment]
Hey! 

Thanks for the feedback on the PR. 

I haven't had a chance to explore Moto filter helpers yet. Could you please point me to the helpers code or implementation?

Added tests as well. 😊

**Comment by viren-nadkarni** [comment]
Hi @ashish1500616, you could take a look at how `describe_vpcs` is implemented in Moto: https://github.com/getmoto/moto/blob/643759bd645f9f9fc5240774fc5186f13bfe625e/moto/ec2/models/vpcs.py#L459-L470

Specifically, check out the `generic_filter` helper.

Also, in our case `security_group_rule_ids` is analogous to `vpc_ids`, if you could adapt a similar flow, that's another bird we can kill.

Do you think we can achieve something minimal and clean like this?

File: localstack/services/ec2/exceptions.py (hunk 1 above)
Comment (viren-nadkarni) [reply localstack/services/ec2/exceptions.py:75]: Looks like a bug in Moto. We should make it as close to AWS behaviour as possible

File: tests/aws/services/ec2/test_ec2.py [hunk 2]
```diff
@@ -415,3 +417,164 @@ def test_pickle_ec2_backend(pickle_backends, aws_client):
...
+        result = aws_client.ec2.describe_security_group_rules(
+            Filters=[
+                {"Name": "group-id", "Values": [security_group["GroupId"]]},
+            ],
+        )
+        assert len(result["SecurityGroupRules"]) == 2
+        for rule in result["SecurityGroupRules"]:
+            if not rule["IsEgress"]:
+                assert rule["Description"] == "UserGroupID Test Description"
+                assert rule["ReferencedGroupInfo"]["UserId"] == DEFAULT_AWS_ACCOUNT_ID
```
Comment (viren-nadkarni) [inline tests/aws/services/ec2/test_ec2.py:538]: `UserId` must be checked against `TEST_AWS_ACCOUNT_ID`

```suggestion
                assert rule["ReferencedGroupInfo"]["UserId"] == TEST_AWS_ACCOUNT_ID
```

**Comment by viren-nadkarni** [comment]
Closing in favour of https://github.com/getmoto/moto/pull/7831

**Pull request closed without merging**